src/natcap/invest/urban_nature_access.py

```diff
@@ -617,63 +816,72 @@ def _greenspace_supply_demand_op(greenspace_budget, population):
-    #  Shouldn't be if we're coming from convolution.
-    out_array = numpy.full(
-        greenspace_area.shape, FLOAT32_NODATA, dtype=numpy.float32)
-
-    # Small negative values should already have been filtered out in another
-    # function after the convolution.
-    # This avoids divide-by-zero errors when taking the ratio.
-    valid_pixels = (convolved_population > 0)
-
-    # R_j is a ratio only calculated for the greenspace pixels.
-    greenspace_pixels = ~numpy.isclose(greenspace_area, 0.0)
-    valid_pixels &= greenspace_pixels
-    if population_nodata is not None:
-        valid_pixels &= ~numpy.isclose(
-            convolved_population, population_nodata)
-
-    if greenspace_nodata is not None:
-        valid_pixels &= ~numpy.isclose(greenspace_area, greenspace_nodata)
-
-    # If the population in the search radius is numerically 0, the model
-    # specifies that the ratio should be set to the greenspace area.
-    population_close_to_zero = numpy.isclose(convolved_population, 0.0)
-    out_array[population_close_to_zero] = (
-        greenspace_pixels[population_close_to_zero])
-    out_array[~greenspace_pixels] = 0.0  # set non-greenspace non-nodata to 0
-    out_array[valid_pixels] = (
-        greenspace_area[valid_pixels] / convolved_population[valid_pixels])
-
-    return out_array
+        Returns:
+            A numpy array with the ratio ``R_j`` representing the
+            greenspace-population ratio with the following constraints:
+
+                * ``convolved_population`` pixels that are numerically close to
+                  ``0`` are snapped to ``0`` to avoid unrealistically small
+                  denominators in the final ratio.
+                * Any non-greenspace pixels will have a value of ``0.0`` in the
+                  output matrix.
+        """
+        # ASSUMPTION: population nodata value is not close to 0.
+        #  Shouldn't be if we're coming from convolution.
+        out_array = numpy.full(
+            greenspace_area.shape, FLOAT32_NODATA, dtype=numpy.float32)
+
+        # Small negative values should already have been filtered out in
+        # another function after the convolution.
+        # This avoids divide-by-zero errors when taking the ratio.
+        valid_pixels = (convolved_population > 0)
+
+        # R_j is a ratio only calculated for the greenspace pixels.
+        greenspace_pixels = ~numpy.isclose(greenspace_area, 0.0)
+        valid_pixels &= greenspace_pixels
+        if population_nodata is not None:
+            valid_pixels &= ~numpy.isclose(
+                convolved_population, population_nodata)
+
+        if greenspace_nodata is not None:
+            valid_pixels &= ~numpy.isclose(greenspace_area, greenspace_nodata)
+
+        # If the population in the search radius is numerically 0, the model
+        # specifies that the ratio should be set to the greenspace area.
+        population_close_to_zero = numpy.isclose(convolved_population, 0.0)
```

Comment:
The default `isclose` tolerances are rtol=1e-05, atol=1e-08. And I think rtol would actually be 0 because the compared value is 0. I wonder if you might want an even higher atol since in the context of this model, 0.00000008 people is probably negligible. Maybe even as much as 0.5 since that would already double the "effective greenspace" (as I'm understanding this ratio).

Reply:
Yeah, that's a really interesting point.  Let me think about this for a bit and I'll get back to you soon.

Reply:
Yep, I agree, I think a higher atol would be a really good idea.

The main reason is that the population raster coming in to this function is coming out of a non-normalized convolution, and so even within float64 space it wouldn't be that hard for the accumulated numerical error to exceed 1e-8.

In practice, I suspect this case will mostly not be triggered, given this is an urban model and that a single low-population pixel would need to be on the very fringes of a search kernel to get such a low number.  I tried running this on the Paris sample data with atols ranging from 1e-8, 1e-7 ... 1e-1, 1 with no detectable differences between the outputs, which isn't surprising for an urban area, even one surrounded by farmland and open space.

Yep, so you're right that if we had a super low population, we'd end up with a higher ratio since there'd be double (in the case of 0.5 people) the greenspace per person.

So in the interest of caution, how about we bump this up to 0.1?

Reply:
In case you agree, I've done this and added a comment in 5275ee0bd